### PR TITLE
Add JSDoc based types

### PR DESCRIPTION
**Pull request by wooorm**
### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 
*   [x] If applicable, I’ve added docs and tests

### Description of changes

This adds types through JSDoc comments.

I’m raising this as a PR because it does not use generics, which are available in the current `.d.ts`, which seem quite useful in `u`. Maybe the JSDoc can be improved, or this is a case where a `.d.ts` file is preferred

### CODE REVIEW

File: index.js
Comment (remcohaszing) [inline index.js]: Removing this line will make tests pass once, then run prettier and break tests on a second run.

This is why I think CI should check code formatting, not reformat the code and then discard those changes. (`prettier --check .` instead of `prettier --write .`.)

I think using `prettier-ignore` should be used very sparingly (I never used it). This case is legit, because otherwise prettier will break code. Since prettier shouldn’t break code, I think an upstream issue should be created and linked here.

File: types/index.d.ts
Comment (remcohaszing) [inline types/index.d.ts:77]: Can we keep type tests when generating types from inline type annotations? Perhaps it makes sense to write test cases using TypeScript. In my experience test code often requires very minimal type annotations.

Comment (wooorm) [reply types/index.d.ts:77]: The tests are also TS through JSDoc, so they are also checking types. What's the benefit of duplicating tests?

File: index.js
Comment (ChristianMurphy) [inline index.js]: Would it be possible to preserve some of the generic type inference from the .d.ts based typings?

File: index.js
Comment (wooorm) [reply index.js]: How is that different from an `eslint-ignore` for example?

---

this seems close to https://github.com/prettier/prettier/issues/6120 btw. But that fix doesn’t apply on functions. [Playground of difference](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEA3AhgJwAQFdsC82A9AFTYACMAngA5zbABmuUYMAltABQCUSeKABM4TDlDhCAvtlLFs3Fm07QFvRlN4AdKDqYQIhEqVJU6DYAEIAYgalyFAIyy8QAGhARaKqAGdkoFiYEADuAApYCP4o6AA2IejU-h6OmOhgANZwMADKtOniAObIMJi4cB5wALaOkiJCADLoUIW46IVwtphV6DCcLcgg6LgwEO4gABYwVbEA6hMc8L75YHA5UYscqIvUg2C+ySDivnCYMGFphT3ITHEnHgBWvgAeAEJpmdk56FVwDeJwG53CogJ7PHJFWJwACKuAg8CBsXuIHymBOmEGzlqsXGtEw4hgsw4QhgE2QAA4AAwePEQE6zNK0QZ4uDo1CAjwARzh8AuXmiQ18AFoJJJJONMHBuRxJRd2tckLckSCTlUOCUyirITCeYDFcCPDB0I4iSSyUgAEyGtIcWJFADCECqCpArIArONcCcACrG6JK5GocoASWECFyYHx3gAgsIcjQoYiTlIpEA)

Comment (remcohaszing) [reply index.js]: ESLint and Prettier have different purposes.

ESLint checks a variety of rules, of which some may not have a proper resolution. For example a project may require imports to be sorted, but if they use a library which requires one import before another, this can’t be reslved. In this case the rule can be disabled using and `eslint-disable-line` comment.

The goal of Prettier is to provide consistent formatting without breaking any code. This `prettier-ignore` comment is necessary, because not having it makes Prettier break the code. Therefor this comment indicates a bug in Prettier.

I’m not saying this comment shouldn’t be here. I’m saying this is a workaround for a bug in Prettier. I think such workarounds should be paired with a comment linking to an upstream issue.

File: types/index.d.ts
Comment (remcohaszing) [reply types/index.d.ts:77]: Basically to ensure this keeps working:

```ts
// $ExpectType { type: "foo" }
const node = u('foo')
```

File: index.js
Comment (remcohaszing) [reply index.js]: I agree this is a nice feature. As an example, the following is currently valid, but would be invalid after merging this PR as-is.

```ts
import { text } from 'mdast'

const text: Text = u('text', '')
```

Comment (ChristianMurphy) [reply index.js]: originally this was supported through overloading, but overloading isn't a part of TS through JSDoc yet https://github.com/microsoft/TypeScript/issues/25590

It might also be possible without overloads using conditional types https://www.typescriptlang.org/docs/handbook/2/conditional-types.html but not :100: sure if it would work and how well that would mix with JSDoc

File: index.js
Comment (wooorm) [reply index.js]: found a way around this: prettier is checking for `@type`, which can also be used

File: index.js
Comment (wooorm) [reply index.js]: I figured out a way to solve this in JSDoc, will push `unist-util-is` with that soon and use the same idea here later

File: index.js
Comment (wooorm) [reply index.js]: Nope...

Reported prettier/prettier#10733.

File: index.js
Comment (wooorm) [reply index.js]: added

File: types/index.d.ts
Comment (wooorm) [reply types/index.d.ts:77]: Added

File: index.test-d.ts
Comment (remcohaszing) [inline index.test-d.ts:26]: I actually like these tests using `tsd` directly instead of `dtslint` even better.

**Review by remcohaszing — APPROVED** [review]

File: package.json
Comment (remcohaszing) [inline package.json:64]: Since formatting may break the build and tests, I think formatting should be run first.

```suggestion
    "test": "npm run format && npm run build && npm run test-coverage"
```

Comment (wooorm) [reply package.json:64]: Reason for doing format after build is that the generated/published types aren't pretty. Running prettier and xo over them ensures they are improved

Comment (remcohaszing) [reply package.json:64]: > Running prettier and xo over them ensures they are improved

I personally agree, but this is really subjective. One of the [goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals) of TypeScript is to output clean code. I imagine this applies to the type definitions as well.

> 4. Emit clean, idiomatic, recognizable JavaScript code.

Anyway, these changes have my approval, but personally I think switching them is better, because it may prevent breakage.

Comment (wooorm) [reply package.json:64]: also experienced it making debugging quite much harder, because the files don’t match the project style if they‘re not formatted. So when inspecting them you first have to dig through tons of formatting issues.

I’d quite want to keep them the order they are, and I’m not 100% on the problems you’re foreseeing by keeping them that way?

**Pull request merged**
